File: AgeOfArenaUnity/Assets/Scripts/ResourceManager.cs
using System;

/// <summary>
/// Per-player resource ledger. Single-player slice = team 0 only. Port of the
/// Three.js ResourceManager, extended with a 4th resource (stone) for the
/// Unity build. Raises <see cref="OnChanged"/> whenever a value changes so the
/// HUD can refresh without polling.
/// </summary>
public class ResourceManager
{
    public int food = 200;
    public int wood = 200;
    public int gold = 100;
    public int stone = 200;  // M8/STONE: AoE2-parity stone economy (Castle/University/towers cost stone).

    public int pop = 0;
    public int popCap = 5;  // TC_BASE_POP

    public event Action OnChanged;

    public int Get(ResourceKind kind) => kind switch
    {
        ResourceKind.Food => food,
        ResourceKind.Wood => wood,
        ResourceKind.Gold => gold,
        ResourceKind.Stone => stone,
        _ => 0,
    };

    public void Gain(ResourceKind kind, int amount)
    {
        if (amount == 0) return;
        switch (kind)
        {
            case ResourceKind.Food: food += amount; break;
            case ResourceKind.Wood: wood += amount; break;
            case ResourceKind.Gold: gold += amount; break;
            case ResourceKind.Stone: stone += amount; break;
        }
        OnChanged?.Invoke();
    }

    public bool CanAfford(int foodCost, int woodCost, int goldCost, int stoneCost)
        => food >= foodCost && wood >= woodCost && gold >= goldCost && stone >= stoneCost;

    public void Deduct(int foodCost, int woodCost, int goldCost, int stoneCost)
    {
        food -= foodCost;
        wood -= woodCost;
        gold -= goldCost;
        stone -= stoneCost;
        OnChanged?.Invoke();
    }

    public void SetPop(int pop, int popCap)
    {
        this.pop = pop;
        this.popCap = popCap;
        OnChanged?.Invoke();
    }
}

File: web/src/core/ResourceManager.ts
/**
 * Per-team resource ledger — ResourceManager.cs port.
 * Raises onChange whenever a value changes so the HUD can refresh without polling.
 */
import { Age, ResourceKind } from "./GameTypes";

export class ResourceManager {
  food = 200;
  wood = 200;
  gold = 100;
  stone = 100;
  pop = 0;
  popCap = 5;
  age: Age = Age.Dark;

  // Research-based gather rate multipliers (start at 1.0, bumped by techs)
  techGatherFoodMult  = 1.0;
  techGatherWoodMult  = 1.0;
  techGatherGoldMult  = 1.0;
  techGatherStoneMult = 1.0;
  // Research-based unit multipliers
  techCavalrySpeedMult    = 1.0;
  techTradeCartSpeedMult  = 1.0;
  /** Supplies: flat food discount on Militia training cost. */
  techMilitiaFoodDiscount = 0;

  onChange: (() => void) | null = null;

  get(kind: ResourceKind): number {
    switch (kind) {
      case ResourceKind.Food:  return this.food;
      case ResourceKind.Wood:  return this.wood;
      case ResourceKind.Gold:  return this.gold;
      case ResourceKind.Stone: return this.stone;
    }
  }

  gain(kind: ResourceKind, amount: number) {
    if (amount === 0) return;
    switch (kind) {
      case ResourceKind.Food:  this.food  = Math.max(0, this.food  + amount); break;
      case ResourceKind.Wood:  this.wood  = Math.max(0, this.wood  + amount); break;
      case ResourceKind.Gold:  this.gold  = Math.max(0, this.gold  + amount); break;
      case ResourceKind.Stone: this.stone = Math.max(0, this.stone + amount); break;
    }
    this.onChange?.();
  }

  canAfford(food: number, wood: number, gold: number, stone = 0): boolean {
    return this.food >= food && this.wood >= wood && this.gold >= gold && this.stone >= stone;
  }

  deduct(food: number, wood: number, gold: number, stone = 0) {
    this.food  = Math.max(0, this.food  - food);
    this.wood  = Math.max(0, this.wood  - wood);
    this.gold  = Math.max(0, this.gold  - gold);
    this.stone = Math.max(0, this.stone - stone);
    this.onChange?.();
  }
}
